source/Ship.cpp

```diff
@@ -4069,3 +4379,346 @@ void Ship::CreateSparks(vector<Visual> &visuals, const Effect *effect, double am
+	bool hasShieldRegen = shieldRegen > 0.;
+	return (hasShieldRegen) ? attributes.Get("shield energy") * (1. + attributes.Get("shield energy multiplier")) : 0.;
+}
+
+
+
+double Ship::GetHullEnergyPerFrame() const
+{
+	double hullRepair = attributes.Get("hull repair rate") * (1. + attributes.Get("hull repair multiplier"));
+	bool hasHullRepair = hullRepair > 0.;
+	return (hasHullRepair) ? attributes.Get("hull energy") * (1. + attributes.Get("hull energy multiplier")) : 0.;
+}
+
+
+
+double Ship::GetRamscoopRegenPerFrame() const
+{
+	if(!currentSystem)
+		return 0.;
+	double scale = GetSolarScale();
+	return currentSystem->SolarWind() * .03 * scale * (sqrt(attributes.Get("ramscoop")) + .05 * scale);
+}
+
+
+
+double Ship::GetRegenEnergyPerFrame() const
+{
+	return GetHullEnergyPerFrame() + GetShieldEnergyPerFrame();
+}
+
+
+
+// Get the current energy.
+double Ship::GetCurrentEnergy() const
+{
+	return attributes.Get("energy capacity") ? energy : (hull > 0.) ? 1. : 0.;
+}
+
+
+
+// Energy loss due to ion storms or ion damage.
+double Ship::GetPotentialIonEnergyLoss() const
+{
+	double ionDamageEnergyCost = 0.;
+	if(ionization)
+	{
+		double ionResistance = attributes.Get("ion resistance");
+		double ionEnergy = attributes.Get("ion resistance energy") ? attributes.Get("ion resistance energy") : 0.;
+		ionEnergy /= ionResistance;
+		ionResistance = .99 * ionization - max(0., .99 * ionization - ionResistance);
+		ionDamageEnergyCost = ionization + ionResistance * ionEnergy;
+	}
+	// Return 60 frames worth of ion energy losses so that fighters can act more quickly to return to carrier.
+	return 60. * ionDamageEnergyCost;
+}
+
+
+
+// Get the total energy consumption per frame combined from idle, moving, firing, shield regen, and hull regen.
+double Ship::GetEnergyConsumptionPerFrame() const
+{
+	return GetIdleEnergyPerFrame() - GetMovingEnergyPerFrame() - GetFiringEnergyPerFrame() - GetRegenEnergyPerFrame();
+}
+
+
+
+double Ship::GetSolarScale() const
+{
+	return .2 + 1.8 / (.001 * position.Length() + 1);
+}
+
+
+
+// Enough energy for 10 seconds of sustained flight is considered minimum
+// energy.  Actual energy minus this value is considered spare energy.
+double Ship::GetSpareEnergy() const
+{
+	double maxEnergy = min(energy, Attributes().Get("energy capacity"));
+	double totalConsumption = 60 * GetEnergyConsumptionPerFrame();
+	double minimumOperationEnergy = 0.;
+	// All energy is available for sharing if there's any energy generation.
+	if(GetIdleEnergyPerFrame() <= 0.)
+	{
+		minimumOperationEnergy = minimumOperatingTime * (-totalConsumption);
+	}
+	return maxEnergy - minimumOperationEnergy;
+}
+
+
+
+// A small check to see if a ship may request help.  Primarily for player escort
+// fleet but can be used by any ship.
+bool Ship::MayRequestHelp() const
+{
+	return (!isTankerCarrier && IsFuelLow()) || IsEnergyLow() || IsDisabled();
```

Comment:
I understand that a tanker carrier shouldnt be requesting the fuel it just gave away.
Just to be sure, we do make sure it keeps a bit of it in reserve for itself to at least jump once? Because IsFuelLow would indicate to me it probably cannot make a jump?

Reply:
The tanker carrier will distribute all fuel except one jump.  It prioritizes refueling the fleet and mission escorts with one jump before giving extra fuel beyond that to the fleet.